### ISSUE TITLE
stub allows code to compile but shouldn't

### ISSUE THREAD

**Issue by xenoterracide** [typetools/checker-framework#4598]
```java
/*
 * Copyright © 2021 Caleb Cushing.
 * Apache 2.0. See https://github.com/xenoterracide/brix/LICENSE
 * https://choosealicense.com/licenses/apache-2.0/#
 */
package com.xenoterracide.brix.configloader.api;

import org.apache.commons.lang3.builder.ToStringBuilder;
import org.apache.commons.lang3.builder.ToStringStyle;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.nio.file.Path;
import java.util.Map;
import java.util.Objects;

public class ProcessedFileConfiguration {

  private final Map<String, @Nullable Object> context;

  private final @Nullable Boolean overwrite;

  private final @Nullable Path source;

  private final Path destination;

  ProcessedFileConfiguration(
    Map<String, @Nullable Object> context,
    @Nullable Boolean overwrite,
    @Nullable Path source,
    Path destination
  ) {
    this.context = context;
    this.overwrite = overwrite;
    this.source = source;
    this.destination = destination;
  }

  public static Builder builder() {
    return new Builder();
  }

  public Map<String, @Nullable Object> getContext() {
    return context;
  }

  public @Nullable Boolean getOverwrite() {
    return overwrite;
  }

  public @Nullable Path getSource() {
    return source;
  }

  @Override
  public String toString() {
    return ToStringBuilder.reflectionToString( this, ToStringStyle.MULTI_LINE_STYLE );
  }

  public Path getDestination() {
    return destination;
  }

  @SuppressWarnings("checkstyle:HiddenField")
  public static final class Builder {
    private Map<String, @Nullable Object> context = Map.of();

    private @Nullable Boolean overwrite;

    private @Nullable Path source;

    private @Nullable Path destination;

    private Builder() {
    }

    public Builder context( Map<String, @Nullable Object> context ) {
      this.context = context;
      return this;
    }

    public Builder overwrite( @Nullable Boolean overwrite ) {
      this.overwrite = overwrite;
      return this;
    }

    public Builder source( @Nullable Path source ) {
      this.source = source;
      return this;
    }

    public Builder destination( Path destination ) {
      this.destination = destination;
      return this;
    }

    public ProcessedFileConfiguration build() {
      Objects.requireNonNull( destination, "destination" );
      return new ProcessedFileConfiguration( context, overwrite, source, destination );
    }
  }
}
```

this code passes with the following `java.util.astub` being present
```java
package java.util;

import org.checkerframework.checker.nullness.qual.Nullable;

public final class Objects {
    public static <T> T requireNonNull( @Nullable T obj);
    public static <T> T requireNonNull( @Nullable T obj, String message);
}
```
of course, it'll do exactly what I mean at runtime, but this doesn't actually guarantee that `destination` is non null, as I'm not actually using the non null return

**Comment by mernst** [comment]
To clarify the issue:
 * Without `-Astubs`, a nullness error is issued for the `Objects.requireNonNull` call, because `destination` might be null.
 * Without `-Astubs` and without the `Objects.requireNonNull` call, a nullness error is issued for the `return new` line, because `destination` might be null.
 * With `-Astubs=java.util.astub`, no error is issued.

This is intended, correct behavior.  Because the stub file contains no `@EnsuresNonNull` annotation, the one in the annotated JDK is used:

```java
    @EnsuresNonNull("#1")
    public static <T> @NonNull T requireNonNull(@NonNull T obj)
```

I have opened a pull request (https://github.com/typetools/checker-framework/pull/4635) to enable declaration annotations in stub files to override declaration annotations in the annotated JDK.  When that is merged, you will be able to write `@EnsuresNonNull({})` in the stub file (see the test case in the pull request).
